File: src/graph_algorithms/depth_first_search.py
def depth_first_search(graph, start, visited=None):
    """
    Perform Depth First Search (DFS) on a graph.

    Parameters:
        graph (dict): A dictionary representing the adjacency list of the graph.
                      Each key is a vertex, and the value is a list of neighbors.
        start (any): The starting vertex for the DFS.
        visited (set): A set to keep track of visited vertices.

    Returns:
        list: A list of vertices in the order they are visited.
    """
    if visited is None:
        visited = set()

    visited.add(start)
    result = [start]

    for neighbor in graph[start]:
        if neighbor not in visited:
            result.extend(depth_first_search(graph, neighbor, visited))

    return result

File: src-tauri/src/Algorithms/graph_algorithms/depth_first_search.rs
pub fn dfs(graph: &Vec<Vec<usize>>, start: usize, visited: &mut Vec<bool>, order: &mut Vec<usize>) {
    visited[start] = true;
    order.push(start);

    for &neighbor in &graph[start] {
        if !visited[neighbor] {
            dfs(graph, neighbor, visited, order);
        }
    }
}

fn main() {
    let graph = vec![vec![1, 2], vec![2], vec![0, 3], vec![3]];

    let start = 2;
    let mut visited = vec![false; graph.len()];
    let mut dfs_order = Vec::new();

    dfs(&graph, start, &mut visited, &mut dfs_order);

    println!("DFS starting from node {}: {:?}", start, dfs_order);
}
